### ISSUE TITLE
Taking the starting tree from a file is still not working as expected.

### ISSUE THREAD

**Issue by Anaphory** [lmaurits/BEASTling#93]
I have just tried to construct an Indo-European analysis using the IE tree from glottolog.
This did not work as easily as expected.

_Excourse: Context_
This was to see whether an issue running an `xml` file generated by Beastling failed – NaNs in all likelihoods – due to an issue with the starting tree. That it failed is a separate issue I will need to raise.
### The expectation, and whether to do anything there

This does require some obvious manipulation of the `tree-glottolog-newick.txt`. 
1. A simple regular expression replaces all the “'Name [glottocode][isocode]'” strings by just the glottocode. – `s/'[^']*[[]\([a-z]\{4\}[0-9]\{4\}\)[]][^']*'/\1/`
2. All non-IE trees need to be removed, because `beastling` looks only in the first tree.
3. Given that languages could be internal nodes with others depending on them (eg. standard language and dialect), translate each internal node into a leaf of that name with a short branch. – `s/)\([a-z]\{4\}[0-9]\{4\}\)/,\1:0.05)/g`

(1) one can't (or shouldn't) be avoided. (2) is hard to do automatically when a file can either contain multiple trees for different families or multiple trees for the same language set (such as a `beast2`-generated `.nex`), so it should remain as it is – and anyway, what if we _did_ assume the file could contain another tree that does contain the target language, and then we get languages in disconnected trees? We don't want that. (3) one does not have an obvious solution that does not introduce nasty length-0 branches into the starting tree, but might be desirable to avoid if we find a way.
## The resulting tree lacks leaf labels.

From this I obtain a tree that Beastling accepts for constructing a starting tree from, but the starting tree constructed is lacking node labels.
[indo_european_with_broken_starting_tree.txt](https://github.com/lmaurits/BEASTling/files/363545/indo_european_with_broken_starting_tree.txt)

**Comment by Anaphory** [comment]
The core issue here is dealing with superset trees. The `starting_tree_tests.py` runs through, but superset-based starting trees are not well formed. I have slightly expanded the `starting_tree_tests.py` to at least check that the leaves of the generated starting tree are as expected, for future testing to immediately die to a problem like this. Now I'll have a look how to fix it – I assume it's our regular problem of having to handle roots and leaves specially when pruning trees.

**Comment by Anaphory** [comment]
I should use a more current version of `newick` – There is an issue with the superset test, but the `newick` module notices its failure because it is tempted to return a non-binary tree.

Also, supersets in general are fine. The thing that is not fine is a superset tree containing siblings of leaves. Singletons are pruned, unused leaves are pruned, but that pruning step can re-introduce singleton leaves due to a snitch in the implementation, I guess – leading to a tree containing all leaf names (different from the bug report above with old newick) but some singletons.

**Comment by lmaurits** [comment]
I'm not sure I 100% follow the details of this issue.  Whenever you get the time, I'd appreciate a high level summary / progress update on this.

**Comment by lmaurits** [comment]
So, it seems like there are two things happening in this issue:
1. You are attempting to take the tree that Glottolog distributes and munge it to suit BEASTling using some combination of regex and the `newick` library and getting unsatisfactory results, perhaps due to bugs in `newick`.
2. BEASTling is actually accepting some of these unsatisfactory results and producing XML files which BEAST will not accept, and you think it should be rejecting them?

Is this about right?  For the sake of 1.2.0, I don't consider 1. above to be BEASTling's problem.  We can work toward making this smoother in the future, but for now I am happy for BEASTling to only accept fully standard Newick trees as starting trees and leave necessary processing of trees to the user.  Problem 2. above _is_ a BEASTling issue, as BEASTling should reject malformed starting trees, including those with unlabelled leaves.

The current `development` code, after processing a user-supplied starting tree (e.g. by pruning taxa not in the linguistic data) now does 3 `assert`s on the resulting tree: that all nodes have exactly 0 or 2 descendants (i.e. there are no redundant internal nodes), that the number of leaves exactly matches the number of languages in the linguistic data, and that every leaf has a label.  It sounds like this should catch the problem trees you describe above?  Please confirm.

**Comment by Anaphory** [comment]
I need a better way to track the data connected to those issues I started. Maybe git tags in my working dir?

From right now, your guess what I wanted here looks very appropriate.

So, I assume that fixes this issue for now, digging through my earlier tries in RomanceEvaluation (I assume that's where I encountered it) has dropped from my agenda after the idea of RomanceEvaluation was replaced by doing stochastic simulations.
